### ISSUE TITLE
多个不同task的ptuning 无法同时加载，一个大模型只能使用一个task的ptuing，无法发挥多个task soft prompt共享大模型的优势

### ISSUE THREAD

**Issue by wolaiye1010** [THUDM/ChatGLM-6B#840]
### Is your feature request related to a problem? Please describe.

多个不同task的ptuning 无法同时加载，一个大模型只能使用一个task的ptuing。

### Solutions

建议可以加载多个，这样大模型的基础模型多个task可以共用，这样就不要new多个大模型，大模型可以共用只需要load一个到显存，每个task也load一份，预测的时候根据不同的task路由到不同的task的ptuning后去prompt再forward公共的大模型

**Comment by summer-silence** [comment]
@wolaiye1010 请问这个有解决方案吗

**Comment by wolaiye1010** [comment]
> @wolaiye1010 请问这个有解决方案吗

已解决，修改chatgpt prompt 模型load 级预测源码，做一个路由，不同的任务路由到不同的sort prompt模型
代码如下：
load修改

`
from transformers import AutoConfig, AutoModel, AutoTokenizer
import torch.nn as nn
import torch
import json
import re
import threading

class ChatModel(nn.Module):
    def __init__(self):
        super(ChatModel, self).__init__()
    
    def loadPrefixEncoder(self,taskType):
        if 1==taskType:
            checkpoint_path='/code/zhudongchang/data/chatglm/finetune-2024045-128-2e-2/checkpoint-3000/pytorch_model.bin'
        elif 2==taskType:
            checkpoint_path='/code/zhudongchang/data/chatglm/finetune-2024045-128-2e-2/checkpoint-3000/pytorch_model.bin'
        else:
            checkpoint_path='/code/zhudongchang/data/chatglm/finetune-2024045-128-2e-2/checkpoint-3000/pytorch_model.bin'
        
        prefixEncoder=PrefixEncoder(self.config)
        prefix_state_dict = torch.load(checkpoint_path)
        new_prefix_state_dict = {}
        for k, v in prefix_state_dict.items():
            if k.startswith("transformer.prefix_encoder."):
                new_prefix_state_dict[k[len("transformer.prefix_encoder."):]] = v
        prefixEncoder.load_state_dict(new_prefix_state_dict)
        prefixEncoder.float()
        prefixEncoder.cuda()
        
        if not hasattr(self.model.transformer, 'prefixEncoder'):
            self.model.transformer.prefixEncoder={}
        self.model.transformer.prefixEncoder[taskType]=prefixEncoder
        
    
    def loadModel(self):
        print('load start')
        # modelPath='/code/zhudongchang/data/chatglm/chatglm-6b-int8'
        modelPath='/code/zhudongchang/data/chatglm/chatGLM-6B'
        self.tokenizer = AutoTokenizer.from_pretrained(modelPath, trust_remote_code=True)
#         model = AutoModel.from_pretrained(modelPath, trust_remote_code=True).half().cuda()

        config = AutoConfig.from_pretrained(modelPath, trust_remote_code=True, pre_seq_len=128)
        self.config=config
        
        model = AutoModel.from_pretrained(modelPath, config=config, trust_remote_code=True)
        self.model=model
        
        self.loadPrefixEncoder(1)
        self.loadPrefixEncoder(2)
        
        model = model.half().cuda()
        model = model.eval()
        print('load end')
        
    
    def chat(self,text):
#         model=model if model is not None else self.model
        textJson=json.loads(text)
        query=textJson[-1]['q']
        
        matchObj = re.match( r'(taskType:(\d+)\s+)(.+)', query,re.S)
        if matchObj:
            query=matchObj.group(3)
            taskType=int(matchObj.group(2))
        else:
            taskType=1
        
        currentThread=threading.current_thread()
        key=f'{currentThread.ident}_{taskType}'
        if not hasattr(self.model.transformer, 'taskTypeList'):
            self.model.transformer.taskTypeList=[]
        self.model.transformer.taskTypeList.append(key)
        
        try:
            max_length=min(max(len(query),2048),20000)
            hisory=[]
            for item in textJson[:-1]:
                hisory.append((item['q'],item['a']))
            response,history = self.model.chat(self.tokenizer,query,history=hisory,max_length=max_length)
            return response
        finally:
            self.model.transformer.taskTypeList.remove(key)
`

预测时前向传播修改如下：

修改 modeling_chatglm.py 文件 get_prompt方法
`

    def get_prompt(self, batch_size, device, dtype=torch.half):
        prefix_tokens = self.prefix_tokens.unsqueeze(0).expand(batch_size, -1).to(device)

        taskTypePrefix=list(filter(lambda x:x.startswith(str(threading.current_thread().ident)),self.taskTypeList))[0]
        taskType=int(taskTypePrefix.split('_')[1])
        prefix_encoder=self.prefixEncoder[taskType]
        # print(f'taskType:{taskType},thread_id:{threading.current_thread().ident},thread_name:{threading.current_thread().name}')

        past_key_values = prefix_encoder(prefix_tokens).type(dtype)
        past_key_values = past_key_values.view(
            batch_size,
            self.pre_seq_len,
            self.num_layers * 2,
            self.num_attention_heads,
            self.hidden_size // self.num_attention_heads
        )
        # seq_len, b, nh, hidden_size
        past_key_values = self.dropout(past_key_values)
        past_key_values = past_key_values.permute([2, 1, 0, 3, 4]).split(2)
        # past_key_values = [(v[0], v[1]) for v in past_key_values]
        return past_key_values
`
使用线程id是为了在线预测时，每个线程一个标识，防止多线程与测时 taskType 2读取到taskType 1 的sort prompt
